Dockerfile

```diff
@@ -1,185 +1,132 @@
+        libglew-dev \
+        qtbase5-dev \
+        libqt5opengl5-dev \
+        libcgal-dev \
+        libceres-dev \
+        python3.10-dev \
+        python3-pip
+
+# Build and install COLMAP.
+RUN git clone https://github.com/colmap/colmap.git && \
     cd colmap && \
+    git checkout "3.9.1" && \
     mkdir build && \
     cd build && \
-    cmake .. -DCUDA_ENABLED=ON \
-             -DCMAKE_CUDA_ARCHITECTURES=${CUDA_ARCHITECTURES} && \
-    make -j `nproc` && \
-    make install && \
-    cd ../.. && \
-    rm -rf colmap
-
-# Create non root user, add it to custom group and setup environment.
-RUN groupadd --gid $USER_GID $USERNAME \
-    && useradd --uid $USER_UID --gid $USER_GID -m $USERNAME -d /home/${USERNAME} --shell /usr/bin/bash 
-# OPTIONAL
-# If sudo privilages are not required comment below line
-# Create simple password for user and add it to sudo group
-# Update group so that it is not required to type password for commands: apt update/upgrade/install/remove
-RUN echo "${USERNAME}:password" | chpasswd \
-    && usermod -aG sudo ${USERNAME} \
-    && echo "%sudo ALL=NOPASSWD:/usr/bin/apt-get update, /usr/bin/apt-get upgrade, /usr/bin/apt-get install, /usr/bin/apt-get remove" >> /etc/sudoers
+    mkdir -p /build && \
+    cmake .. -GNinja "-DCMAKE_CUDA_ARCHITECTURES=${CUDA_ARCHITECTURES}" \
+        -DCMAKE_INSTALL_PREFIX=/build/colmap && \
+    ninja install -j1 && \
+    cd ~
+
+# Upgrade pip and install dependencies.
+# pip install torch==2.2.2 torchvision==0.17.2 --index-url https://download.pytorch.org/whl/cu118 && \
+RUN pip install --no-cache-dir --upgrade pip 'setuptools<70.0.0' && \
+    pip install --no-cache-dir torch==2.1.2+cu118 torchvision==0.16.2+cu118 'numpy<2.0.0' --extra-index-url https://download.pytorch.org/whl/cu118 && \
+    git clone --branch master --recursive https://github.com/cvg/Hierarchical-Localization.git /opt/hloc && \
+    cd /opt/hloc && git checkout v1.4 && python3.10 -m pip install --no-cache-dir . && cd ~ && \
+    TCNN_CUDA_ARCHITECTURES="${CUDA_ARCHITECTURES}" pip install --no-cache-dir "git+https://github.com/NVlabs/tiny-cuda-nn.git@b3473c81396fe927293bdfd5a6be32df8769927c#subdirectory=bindings/torch" && \
+    pip install --no-cache-dir pycolmap==0.6.1 pyceres==2.1 omegaconf==2.3.0
+
+# Install gsplat and nerfstudio.
+# NOTE: both are installed jointly in order to prevent docker cache with latest
+# gsplat version (we do not expliticly specify the commit hash).
+#
+# We set MAX_JOBS to reduce resource usage for GH actions:
+# - https://github.com/nerfstudio-project/gsplat/blob/db444b904976d6e01e79b736dd89a1070b0ee1d0/setup.py#L13-L23
+COPY --from=source /tmp/nerfstudio/ /tmp/nerfstudio
+RUN export TORCH_CUDA_ARCH_LIST="$(echo "$CUDA_ARCHITECTURES" | tr ';' '\n' | awk '$0 > 70 {print substr($0,1,1)"."substr($0,2)}' | tr '\n' ' ' | sed 's/ $//')" && \
+    export MAX_JOBS=4 && \
+    pip install --no-cache-dir git+https://github.com/nerfstudio-project/gsplat.git && \
```

Comment:
```suggestion
    pip install --no-cache-dir git+https://github.com/nerfstudio-project/gsplat.git@v1.0.0 && \
```

Reply:
Hi, 
I added the above suggestion.
The latest main branch version of gsplat seems to have some problems with nerfstudio splatfacto.

To fix it, use a pinned version of gsplat, which is known to work. 
v1.0.0 is the version listed in the pyproject.toml file, so suggest using that.

Alternatively, the gsplat tag/branch could be stored in an ARG, as GSPLAT_VERSION. This  would enable it to be overridden during docker build.

Reply:
I guess we should allow gsplat version to be specified as arg, the question is what default to choose. I'm a bit afraid people forget updating this dockerfile and we want to use recent versions of gsplat. I was thinking of installing last released version, but not sure. What do you think?

Reply:
Yes, an ARG would be good, so people can choose to override in their own docker builds.

IMO, the default gsplat version in the docker image should be a stable version of gsplat, which follows whatever gsplat version the pip install nerfstudio versions use.  The version specified in pyproject.toml is 1.0.0.

See the reason the gsplat version was pinned in the pyproject.toml in 
https://github.com/nerfstudio-project/nerfstudio/pull/3200

Reply:
Is it possible to just take the pinned version from the `pyproject.toml`?

Reply:
Maybe just drop the `pip install ... gsplat.git` command, and only use the next command `pip install /tmp/nerfstudio `? That should make nerfstudio use the gsplat 1.0.0 as specified in pyproject.toml. This should be the preferred, stable version.

Nerfstudio splatfacto presumably needs to be updated to be compatible with the latest gsplat version at any point in time. But contributors working on splatfacto should comment on that.

Reply:
^this makes sense to me!

Reply:
After some testing without the pip install gsplat command, it looks like a gsplat build of some version is needed after all. 
The gsplat wheel needs to be produced to be included in the docker image.
Without the prebuilt wheel, the docker image will contain just the source files. 
And as the final docker image will be without builds tooling such as nvcc, the wheel can not be built at runtime / JIT, which makes splatfacto runs fail with "No CUDA toolkit found. gsplat will be disabled" from https://github.com/nerfstudio-project/gsplat/blob/main/gsplat/cuda/_backend.py#L141.

Reply:
Also, with the current command sequence, a pinned gsplat v1.0.0 build of is needed, as the "pip install nerfstudio" will use the gsplat 1.0.0, and would uninstall any newer or older versions (including the wheel), and replace with a non-compiled version of 1.0.0.

So, my original suggestion "pip install --no-cache-dir git+https://github.com/nerfstudio-project/gsplat.git@v1.0.0" is the only way I have been able to successfully run splatfacto in the docker image.

Reply:
(1)

For just relying on the nerfstudio dependencies, would it help to trigger the gsplat JIT compilation via:
```sh
python -c "from gsplat.cuda import _backend"
```
after the pip install? cc @liruilong940607

This may or may not require https://github.com/nerfstudio-project/gsplat/pull/319 to make it into a release.

(2)

Alternatively, we can do some grep magic with the nerfstudio pyproject.toml?
```
GSPLAT_VERSION=$(grep -oP 'gsplat==\K[^"]+' pyproject.toml)
pip install --no-cache-dir git+https://github.com/nerfstudio-project/gsplat.git@v$GSPLAT_VERSION
```

Reply:
We could also wait for https://github.com/nerfstudio-project/gsplat/pull/365

Reply:
I think option 1 is nice. I think I had something similar at one point and later switched to latest build (I thing somebody wanted it but I don’t remember).